### PR TITLE
Add missing Get / List, Create, and Delete methods to clusterclient. Add a wrapper Delete method for Kubectl.

### PR DESCRIPTION
**Pull request by spew**
**What this PR does / why we need it**:
This PR adds more methods to clusterclient to flesh out the things that will be needed for delete. It is submitted separate from delete to make that PR smaller. If it is necessary to see delete in its entirety to reason about this PR, I am happy to open up a complete PR with all of the changes.

**Special notes for your reviewer**:
Note that this PR copies the style and flow of the existing clusterclient. For example, it adds relevant wait code to to the deletion methods, copying what it is done in existing Create methods like CreateMachineObjects.

You will notice, there is a lot of duplicate waiter code. I did attempt to generalize the wait and then call that generalized function. However, I was unable to come up with anything better than copy & paste because the "Items" member of the various cluster API list responses is not a method and therefore can't be used in an interface.

**Release note**:
```release-note
NONE
```

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

File: clusterctl/clusterdeployer/clusterclient.go [hunk 1]
```diff
@@ -129,16 +161,42 @@ func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
...
+	if err != nil {
+		return fmt.Errorf("error creating cluster: %v", err)
+	}
 	return err
 }
 
+func (c *clusterClient) CreateMachineDeploymentObjects(deployments []*clusterv1.MachineDeployment) error {
+	// TODO: Support specific namespaces https://github.com/kubernetes-sigs/cluster-api/issues/252
+	for _, deploy := range deployments {
+		// TODO: Run in parallel https://github.com/kubernetes-sigs/cluster-api/issues/258
```
Comment (k4leung4) [inline clusterctl/clusterdeployer/clusterclient.go]: the issue reference is for machines, do you want to update the issue to include machine deployments and machinesets or file a separate issue?

Comment (spew) [reply clusterctl/clusterdeployer/clusterclient.go]: Added a comment on the issue (I can't edit the original text) that MachineSets and MachineDeployments should be included as well.

File: clusterctl/clusterdeployer/clusterclient.go [hunk 2]
```diff
@@ -129,16 +161,42 @@ func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
...
+		_, err := c.clientSet.ClusterV1alpha1().MachineDeployments(apiv1.NamespaceDefault).Create(deploy)
+		if err != nil {
+			return fmt.Errorf("error creating a machine deployment object: %v", err)
+		}
+	}
+	return nil
+}
+
+func (c *clusterClient) CreateMachineSetObjects(machineSets []*clusterv1.MachineSet) error {
```
[the remark this answers is not in the document]
Comment (spew) [reply clusterctl/clusterdeployer/clusterclient.go]: Added.

File: clusterctl/clusterdeployer/clusterclient.go [hunk 3]
```diff
@@ -129,16 +161,42 @@ func (c *clusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
...
+	for _, ms := range machineSets {
+		// TODO: Run in parallel https://github.com/kubernetes-sigs/cluster-api/issues/258
+		_, err := c.clientSet.ClusterV1alpha1().MachineSets(apiv1.NamespaceDefault).Create(ms)
+		if err != nil {
+			return fmt.Errorf("error creating a machine set object: %v", err)
+		}
```
[the remark this answers is not in the document]
Comment (spew) [reply clusterctl/clusterdeployer/clusterclient.go]: I did consider it, and I think we *should* for consistency's sake. However, I don't think all clients would want to wait for all machines underlying a machineset to be created, etc. 

As we discussed in person, waiting after every operation is too heavy handed as not all clients need to wait. We think the best thing to do is refactor this waiting logic out of this file and into a waiting layer that the clusterdeployer can use as needed. Tagging @jessicaochen so she understands why this change is made if it is done before she returns. 

I created an issue here for this refactor which I will prioritize immediately after delete is finished: https://github.com/kubernetes-sigs/cluster-api/issues/402

File: clusterctl/clusterdeployer/clusterclient.go [hunk 4]
```diff
@@ -148,6 +206,61 @@ func (c *clusterClient) CreateMachineObjects(machines []*clusterv1.Machine) erro
...
+	if err != nil {
+		return fmt.Errorf("error waiting for machine set(s) deletion to complete: %v", err)
+	}
+	return nil
+}
+
+func (c *clusterClient) DeleteMachineObjects() error {
+	err := c.clientSet.ClusterV1alpha1().Machines(apiv1.NamespaceDefault).DeleteCollection(newDeleteOptions(), metav1.ListOptions{})
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (spew) [reply clusterctl/clusterdeployer/clusterclient.go]: I was thinking that the caller is already doing htat (clusterdeployer.go) but I think it should be done here instead. Adding.

**Comment by spew** [comment]
Updated after the last round of feedback.

**Comment by k4leung4** [comment]
lgtm

File: clusterctl/clusterdeployer/clusterclient.go [hunk 5]
```diff
@@ -171,8 +285,72 @@ func (c *clusterClient) WaitForClusterV1alpha1Ready() error {
...
+		return true, nil
+	})
+}
+
+func (c *clusterClient) waitForMachinesDelete() error {
+	return util.PollImmediate(RetryIntervalResourceDelete, TimeoutResourceDelete, func() (bool, error) {
+		glog.V(2).Infof("Waiting for machine objects to be deleted...")
+		response, err := c.clientSet.ClusterV1alpha1().Machines(apiv1.NamespaceDefault).List(metav1.ListOptions{})
+		if err != nil {
+			return false, nil
```
Comment (mkjelland) [inline clusterctl/clusterdeployer/clusterclient.go]: Sorry, I missed this on the first round, shouldn't `err` should be returned here? Same thing in the other wait for delete methods.

Comment (spew) [reply clusterctl/clusterdeployer/clusterclient.go]: err should not be returned, from the PollImmediate docs: https://godoc.org/k8s.io/apimachinery/pkg/util/wait#PollImmediate

`PollImmediate tries a condition func until it returns true, an error, or the timeout is reached.`

This means that if err is returned the polling will cease.

Comment (mkjelland) [reply clusterctl/clusterdeployer/clusterclient.go]: Actually, nevermind. I see other examples of PollImmediate being used where the error is not returned.

Comment (mkjelland) [reply clusterctl/clusterdeployer/clusterclient.go]: Oh you responded before me! Yes, you are right :)

**Pull request merged**